File: 3. Behavioral Design Patterns/State/Kotlin/StateTemplate.kt
// State Pattern – Kotlin Template

interface State { fun handle(context: Context) }
class Context(private var state: State) { fun setState(s: State){ state = s }; fun request(){ state.handle(this) } }
class ConcreteStateA : State { override fun handle(c: Context){ /* work */ c.setState(ConcreteStateB()) } }
class ConcreteStateB : State { override fun handle(c: Context){ /* work */ c.setState(ConcreteStateA()) } }
fun main(){ val c = Context(ConcreteStateA()); c.request(); c.request() }

File: 3. Behavioral Design Patterns/State/Typescript/StateTemplate.ts
// State Pattern – TypeScript Template

interface State { handle(context: Context): void }
class Context { constructor(private state: State){} setState(s: State){ this.state = s } request(){ this.state.handle(this) } }
class ConcreteStateA implements State { handle(c: Context){ /* work */ c.setState(new ConcreteStateB()) } }
class ConcreteStateB implements State { handle(c: Context){ /* work */ c.setState(new ConcreteStateA()) } }
;(function main(){ const c = new Context(new ConcreteStateA()); c.request(); c.request() })()

export {}
